Nexus validation tutorial (#402)
FAIRmat-NFDI/pynxtools: docs/how-tos/validate-nexus-file.md

# Validating NeXus files

Note: This is a how-to guide for using different tools to validate NeXus files. If you want to learn more about how validation is done in `pynxtools`, please visit the [explanation page](../learn/nexus-validation.md).

> # The goal

Author: yea, changed it accordingly. Ill check it with mkdocs serve

> # The goal

Author: reworked all these in latest commit

Use a tool to validate NeXus files to a given set of NeXus definitions:

  1. [FAIRmat](https://fairmat-nfdi.github.io/nexus_definitions/index.html#)

  2. [NIAC](https://manual.nexusformat.org/)

# Validation of a .nxs file

The validity of NeXus files is fundamental to ensure FAIR data. Without specific requirements, it is not possible to understand the data. What type of experiment? What Laser Wavelength? Which voltage? What data is represented at all in the table? What is the unit of the value? Which ISO norm does this refer to? Where was this measured? Which year was this measured?

> Therefore you have enter all required fields in the NeXus definition. The requirements are set by the community via workshops or at conferences. You can as well comment the NeXus definitions, to initiate or propose changes/additions. Go to the NeXus definition, and sign-up/log-in and give us some feedback (Red boxes in the image. Expand this panel on the left by clicking on the arrow symbol).  

Reviewer: As suggestions, feel free to change:
```suggestion
The NeXus application definitions define the minimum set of terms that must be used in an instance of that class (i.e., the required terms that you must add to the file in order to be compliant with that application definition). Application definitions also may define terms that are optional in the NeXus data file. The requirements are set by the community via workshops or at conferences. To initiate or propose changes/additions, you can comment the FAIRMat NeXus proposal by going to the [NeXus definitions](https://fairmat-nfdi.github.io/nexus_definitions/index.html#), and using the hypothes.is tool (sign-up/log-in) to give us some feedback (Red boxes in the image. Expand this panel on the left by clicking on the arrow symbol).  
```

![image.png](<./attachments/9d748230d54a8059-image.png>)

> Oftentimes, there will be errors in a generated NeXus file (be it by hand or automatically): Typos, missing required concepts, missing attributes, using the incorrect datatype or format (e.g., array instad of list, float instead of integer, etc.). Therefore, a validation is required, to ensure that the data you want to share, is FAIR.

Reviewer: ```suggestion
Oftentimes, there will be errors in a generated NeXus file (be it by hand or automatically): Typos, missing required concepts, missing attributes, using the incorrect datatype or format (e.g., array instad of list, float instead of integer, etc.). Therefore, a validation is required, to ensure that the data you want to share is FAIR.
```

The NeXus file is valid if it complies with the respective NeXus application definition.

This validation is done by software.

> # Validation software

Reviewer: see above

There are right now three tools, which can be used for validation of NeXus files. All are different and have individual advantages or disadvantages:

1. pynxtools 

2. cnxvalidate

3. punx

Open software is usually shared on Github - There you find usually the most accurate information, as documentation sometimes lags behind. There you see a box with folders and files. Below is the content of the README.md file displayed. This usually shows instructions for installation and handling of the software.

> Here are the GitHub links for the thee software packages:

Reviewer: ```suggestion
Here are the GitHub links for the three software packages:
```
```suggestion
Here are the GitHub links for the thee software packages:
```

> [pynxtools](<https://github.com/FAIRmat-NFDI/pynxtools>)

Reviewer: I switched the order here to reflect the order above

Author: thanks

[cnxvalidate](<https://github.com/nexusformat/cnxvalidate>)

[punx](<https://github.com/prjemian/punx>)

In the following, each package and its capabilities is presented.

### Operating systems

Almost all PC users are used to Windows as operating system.

A lot of software development is done on Linux as operating system.

This is not a problem for big companies, but for smaller open software projects, which are often developed without funding, this is a problem.

If you are used to Windows, consider setting up a Linux operating system to eliminate problems in the installation process and ensure compatibility.

> # 1. pynxtools - Python Nexus Tools

Reviewer: These numbers should be the same as above - again, use markdown syntax for sub-headings

> 

Reviewer: Add here a note to say that the validation in pynxtools is described in `../learn/nexus-validation.md`

Author: ill added a note.

This is a python package which is developed by the FAIRmat consortium.

As a python package, this can be used on Linux and Windows systems.

The package can be installed via pip. Therefore you need to have installed:

1. python

2. pip

> This tool has 3 command line functions:
> 
> 1. [dataconverter](<https://github.com/FAIRmat-NFDI/pynxtools/blob/master/src/pynxtools/dataconverter/README.md>)
> 
> 2. [read\_nexus](<https://github.com/FAIRmat-NFDI/pynxtools/blob/master/src/pynxtools/nexus/README.md>)
> 
> 3. [generate\_eln](<https://github.com/FAIRmat-NFDI/pynxtools/blob/master/src/pynxtools/eln_mapper/README.md>)
> 
> 4. [verify\_nexus](https://github.com/FAIRmat-NFDI/pynxtools/pull/333)

Reviewer: ```suggestion
```
We should not put these here, as this is changing (actually, you are naming 4 tools here already).

Author: yea, right.

> For validation purposes, we will use the "read\_nexus" and "verify\_nexus" function.

Reviewer: ```suggestion
For validation purposes, we will use the "read\_nexus" and "verify\_nexus" command line tools from `pynxtools`.
```

# 1.1 verify_nexus

This tool is currently in development. It enables a command like:

```
verify_nexus C:\nexusvalidation\Raman.nxs
```

The output warning looks like this:
```
...
WARNING: Field /entry/instrument/beam_incident/wavelength/@units written without documentation.
...
```

# 1.1.1 Installation of verify_nexus

> Aside from read_nexus, there is a second functionality from pynxtools. This method is currently in development (Aug 2024). This function is called: verify\_nexus. Therefore, you have to use the development install, until this function is published.
> 

Reviewer: You have not mentioned read_nexus so far, so doesn't make sense to talk about a second functionalit
```suggestion
The `verify_nexus` function is currently under development (Aug 2024). Therefore, you have to install pynxtools from its feature branch until this function is published.
```

Do this to install pynxtools with verify\_nexus

```
git clone https://github.com/FAIRmat-NFDI/pynxtools/tree/hdf-based-validation
cd pynxtools
git checkout git checkout hdf-based-validation
git submodule sync --recursive
git submodule update --init --recursive --jobs=4
python -m pip install --upgrade pip
python -m pip install -e .
python -m pip install -e ".[dev]"

> ```

Reviewer: can be replaced by 
```console
pip install git+https://github.com/AIRmat-NFDI/pynxtools@hdf-based-validation
```

Author: okay, interesting to know this. Ill test that out and replace it then

> then, the help function should be available:

Reviewer: ```suggestion
Then, you should be able to call its help function:
```

```
verify_nexus --help
```

with this output:

```
Usage: verify_nexus [OPTIONS] FILE

  Verifies a nexus file

Options:
  --help  Show this message and exit.
```

# 1.1.2 Using verify_nexus

Open your terminal. Assuming there is a folder at:

For Linux:

```
/home/USER/nexusvalidation
```

For Windows:

```
C:\nexusvalidation
```

Put into this folder your NeXus file, for example this [Raman.nxs file](https://zenodo.org/records/13373909/files/Raman.nxs?download=1).

Use verify_nexus with the command:

```
verify_nexus C:\nexusvalidation\Raman.nxs
```

The respective output is:

```
WARNING: Field /entry/data/spectrum_data_x/@units written without documentation.
WARNING: Field /entry/data/spectrum_data_x_Raman/@units written without documentation.
WARNING: Field /entry/data/spectrum_data_y/@units written without documentation.
WARNING: Field /entry/instrument/beam_incident/wavelength/@units written without documentation.
WARNING: Field /entry/instrument/detector_DU970BV/number_of_cycles/@units written without documentation.
Invalid: The entry `entry` in file `Raman.nxs` is NOT a valid file according to the `NXraman` application definition.
```

# 1.2 read_nexus

> 

Reviewer: Explain here why there's two tools at all:
```suggestion
While `verify_nexus´ is used as a straightforward tool for validating a NeXus file, `read_nexus` outputs a debug log for a given NeXus file by annotating the data and metadata entries with the schema definitions from the respective NeXus base classes and application definitions to which the file refers to. This can be helpful to extract documentation and understand the concept defined in the NeXus application definition.
```

value: [ 288.5499878  289.         288.4500122 ... 1875.        1889.349976 ...
DEBUG: value: [ 288.5499878  289.         288.4500122 ... 1875.        1889.349976 ...
Dataset referenced as NXdata SIGNAL
DEBUG: Dataset referenced as NXdata SIGNAL
===== ATTRS (//entry/data/spectrum_data_y@long_name)
DEBUG: ===== ATTRS (//entry/data/spectrum_data_y@long_name)
value: Raman Intensity 
DEBUG: value: Raman Intensity 
Dataset referenced as NXdata SIGNAL
DEBUG: Dataset referenced as NXdata SIGNAL
===== ATTRS (//entry/data/spectrum_data_y@units)
DEBUG: ===== ATTRS (//entry/data/spectrum_data_y@units)
value: counts 
DEBUG: value: counts 

DEBUG: 
For Axis #0, 1 axes have been identified: [<HDF5 dataset "spectrum_data_x_Raman": shape (1600,), type "<f8">]
DEBUG: For Axis #0, 1 axes have been identified: [<HDF5 dataset "spectrum_data_x_Raman": shape (1600,), type "<f8">]
```

Search for fields which are not found in the NeXus definition by searching for the line: "DEBUG: NOT IN SCHEMA". Recheck the used NeXus definition to eliminate the problem. Be careful with upper and lower case notation and correct spelling.

> Keep in mind, that the output provides quite some information. This is useful for software development, but may be a bit too much for validation purposes.

Reviewer: ```suggestion
Keep in mind that the output provides quite some information. This is useful for software development, but may be a bit too much for validation purposes.
```

> Similar features as the tables and messages provided from punx and cnxvalidate for pynxtools are [planned to be implemented in the future](<https://github.com/FAIRmat-NFDI/pynxtools/pull/333>).

Reviewer: Maybe remove? We have this covered above.

Author: yea, will remove it. Was from my first draft

Author: removed

# 2. cnxvalidate

This package is written in C. It is allows a command line evocation like:

```
nxvalidate -l appdefdir datafile
```

1. nxvalidate: calls the software function

2. \-l appdefdir: points to the location of the NeXus definitions you want to use. This is a path to a folder called "defintions".

3. datafile: This is the path to the .nxs file which should be checked.

This output shows warnings like:

```
definition=NXoptical_spectroscopy.nxdl.xml message="Required attribute URL missing" nxdlPath=/NXentry/definition sev=error dataPath=/entry/definition dataFile=NXopt_minimal_example.nxs
```

and indicates the entry of the .nxs file, which is incorrect and what the respective problem is. It also points to the NeXus definition (.nxdl.xml file), in which this conflict was found.

# 2.1. Installation (Linux only)

> This [did not work for me on windows](/installation_notes_nxvalidate.md) (The problem was the software cmake couldn't fine the libxml2 library. Though, if you solve this, this maybe work on windows).

Reviewer: Remove the personal reference "me".

Maybe:
"Note: You can find more information about installing nxvalidate [here](installation_notes_nxvalidate.md). Note that installation on Windows can be tricky because cmake can sometimes not find the libxml2 library. Though, if you solve this, this maybe work on windows)."

Author: used it as suggested.

Therefore: Use linux.

> The installation process has to be build from source. This is eased significantly by using another software called: cmake.

Reviewer: ```suggestion
The software has to be built from source. This is eased significantly by using another software called: cmake.
```

### Install cmake, github, hdf5 & xml2 library, etc:

Open the terminal and install all parts required to install cnxvalidate via cmake:

```
sudo apt-get update
sudo apt-get install git
sudo apt-get install build-essential
sudo add-apt-repository universe
sudo apt-get install libhdf5-serial-dev
sudo apt-get -y install pkg-config     
sudo apt upgrade -y
sudo apt-get -y install cmake
sudo apt-get install libxml2-dev
```

> ### Directoty location

Reviewer: ```suggestion
### Directory location
```

> create a folder named "nexusvalidation" via terminal or file manager.

Reviewer: ```suggestion
Create a folder named "nexusvalidation" via terminal or file manager.
```

The folder is located at `/home/USER/nexusvalidation`

"USER" is your user name. (You can get your username by the terminal command: `echo $USER`)

In the terminal, this is indicated by `~/nexusvalidation` ( \~ = /home/USER)

> open the thermal and go into this directory by:

Reviewer: ```suggestion
Open the thermal and go into this directory by:
```

Author: used as suggested

```
cd /home/USER/nexusvalidation
```

### Using GitHub

Go to the [Github Repository of cnxvalidate:](<https://github.com/nexusformat/cnxvalidate>)

Click on the green "<> Code" button.

Click on "HTTPS".

Copy the https link.

![image.png](<./attachments/e28ec15bb537c9b7-image.png>)

> open the terminal and ensure you are in the `nexusvalidation` folder.
> 
> clone the github repository (= download the files of the software).

Reviewer: ```suggestion
Open the terminal and ensure you are in the `nexusvalidation` folder.

Clone the github repository (= download the files of the software).
```
suggestion to make each line start either capizalized or not, but consistenly so (also below)

```
cmake ../
```

install cnxvalidate after it was sucessfully build

```
make
```

Now the above mentioned commands should be avaialble. The programm/executable is located at:

```
/home/USER/nexusvalidation/cnxvalidate/build/nxvalidate
```

# 2.2. Using cnxvalidate

Now you can start to validate your created [NeXus file](https://zenodo.org/records/13373909). But before the validation, we need to get a set of NeXus definitions, which we want to use as reference. This is done again by using git:

### Getting NeXus definitions

go to the folder nexusvalidation

```
cd /home/USER/nexusvalidation
```

Download a set of NeXus definitions. Choose only one:

For FAIRmat NeXus definitions, go to [this link](<https://github.com/FAIRmat-NFDI/nexus_definitions>) and copy the github "Code" line to clone the repository. Then:

```
git clone https://github.com/FAIRmat-NFDI/nexus_definitions.git
```

For the NIAC NeXus definitions, go to [this link](<https://github.com/nexusformat/definitions>) and copy the github "Code" line to clone the repository. Then:

```
git clone https://github.com/nexusformat/definitions.git
```

> Now you have a folder called "definitions" in the "nexusvalidation" folder. The path to this definitions folder is used as option for cnxvalidate, to tell the program, which NeXus definitions shall be used.

Reviewer: ```suggestion
Now you have a folder called "definitions" in the "nexusvalidation" folder. The path to this definitions folder is used as option for cnxvalidate, to tell the program which NeXus definitions shall be used.
```

Author: used as suggested

definition=NXoptical_spectroscopy.nxdl.xml message="Required group missing" nxdlPath=/NXentry/NXdata sev=error dataPath=/entry dataFile=NXopt_minimal_example.nxs 
9 errors and 11 warnings found when validating NXopt_minimal_example.nxs
```

The errors tell you now which things are missing (message="Required group missing"), if there is a field missing (message="Required field missing"), or if an attribute is missing (message="Required attribute URL missing" - here for example the attribute named URL)

Now adjust the file creation, and add the respective fields to make your NeXus file compliant with the NeXus definitions. This way, you can ensure that your data is FAIR, which is then ready for sharing and publication.

# 3. Punx - Python Utilities for NeXus HDF5 files

This is python package, and can therefore be used on Linux and Windows systems.

The package can be installed via pip. Therefore you need to have installed:

1. python

2. pip

You can then evoke a command like this:

```
punx validate [-h] [--report REPORT] infile
```

> "validate" is tells the program, that we want to validate a file

Reviewer: ```suggestion
"validate" tells the program that we want to validate a file
```

Author: used as suggested

"[-h]" tells the program to show the help message

> "[--report REPORT]" tells the program, what findings should be reported.  

Reviewer: ```suggestion
"[--report REPORT]" tells the program what findings should be reported.  
```

======= ====== ========== ======================================
```

can be ignored and is a bug right now. If this is the only Error message, then your NeXus file is compliant with the NeXus definitions and you can share and publish your data.

# 3.4. Further Notes

1. Punx only uses the NeXus definiton from the NIAC [NeXus definiton from the NIAC](<https://manual.nexusformat.org/>). The use of the [FAIRmat NeXus definition](<https://fairmat-nfdi.github.io/nexus_definitions/index.html#>) is not possible right now.

2. [Other punx commands are availble](<https://punx.readthedocs.io/en/latest/overview.html#>)

3. [More details for installation](<https://punx.readthedocs.io/en/latest/install.html>)

4. [Github project](<https://github.com/prjemian/punx>)

# Summary

This tutorial showed:

1. 3 different tools for NeXus file validation

2. How to install these tools

3. How to use them via Examples

### Recommended methods:

As pynxtools verify_nexus method is right now in development, [not all situations are covered right now](/testing-validation-tools.md). Therefore, the most reliable method right now is a combination of _Human Manual Validation_ + _Software solutions_.

> ### Pynxtools Parsers:

Reviewer: Either remove this part or, if you want to keep it, simply make a link to ../reference/plugins.md (in order to keep single source principle).

Author: yea, put in a reference and just kept the line 767.

Author: rechecked, is included

For a specifically structured set of data, a parser can be written, which uses the meta data and a pre-structured meta data file, to create a NeXus file. Tough, the parser depends on: Experimental Technique and Setup and has therefore to be written individually. This is another functionallity of [pynxtools with plugins for the techniques](https://fairmat-nfdi.github.io/pynxtools/reference/plugins.html):

[electron microscopy (EM)](<https://github.com/FAIRmat-NFDI/pynxtools-em>)

[x-ray photoelectron spectroscopy (XPS)](<https://github.com/FAIRmat-NFDI/pynxtools-xps>)

[scanning tunneling spectroscopy/microscopy and atomic force microscopy (STS / STM /AFM)](<https://github.com/FAIRmat-NFDI/pynxtools-stm>)

[x-ray diffraction (XRD)](<https://github.com/FAIRmat-NFDI/pynxtools-xrd>)

[ellipsometry (ellips)](<https://github.com/FAIRmat-NFDI/pynxtools-ellips>)

[Raman spectroscopy (raman)](<https://github.com/FAIRmat-NFDI/pynxtools-raman>)

[atom probe microscopy (APM)](<https://github.com/FAIRmat-NFDI/pynxtools-apm>)

### Feedback and contact:

> 1. Best way is to contact the software development directly via a [Github Issue](https://github.com/FAIRmat-NFDI/pynxtools/issues/new).

Reviewer: ```suggestion
1. Best way is to contact the software development team directly via a [Github Issue](https://github.com/FAIRmat-NFDI/pynxtools/issues/new).
```

Author: is already included
